### PR TITLE
[WiP] Fix up existing protocols

### LINKED ISSUE #1: Some accessions have duplicates of dicoms
Since originally sub-directories did not include sequence number as a prefix:

```shell
smaug:/mnt/btrfs/dbic/inbox/DICOM/2016/10
$> ls */A000033 -l
total 0
drwxr-xr-x 1 yoh yoh 15360 Oct 18 02:47 01-anat-scout/
drwxr-xr-x 1 yoh yoh   600 Oct 18 02:47 02-anat-scout_MPR_sag/
drwxr-xr-x 1 yoh yoh   360 Oct 18 02:47 03-anat-scout_MPR_cor/
drwxr-xr-x 1 yoh yoh   360 Oct 18 02:47 04-anat-scout_MPR_tra/
drwxr-xr-x 1 yoh yoh 23040 Oct 18 02:46 05-anat-T1w/
drwxr-xr-x 1 yoh yoh 15360 Oct 18 02:55 06-fmap_run+/
drwxr-xr-x 1 yoh yoh  7680 Oct 18 02:55 07-fmap_run+/
drwxr-xr-x 1 yoh yoh 33600 Oct 18 02:41 08-func_run\=_task-1back/
drwxr-xr-x 1 yoh yoh 33600 Oct 18 02:44 09-func_run+_task-1back/
drwxr-xr-x 1 yoh yoh 33600 Oct 18 02:53 10-func_run+_task-oddball/
drwxr-xr-x 1 yoh yoh 33600 Oct 18 02:53 11-func_run+_task-oddball/
drwxr-xr-x 1 yoh yoh 23040 Oct 14 14:07 anat-T1w/
drwxr-xr-x 1 yoh yoh 15360 Oct 14 14:07 anat-scout/
drwxr-xr-x 1 yoh yoh   360 Oct 14 14:07 anat-scout_MPR_cor/
drwxr-xr-x 1 yoh yoh   600 Oct 14 14:07 anat-scout_MPR_sag/
drwxr-xr-x 1 yoh yoh   360 Oct 14 14:07 anat-scout_MPR_tra/
drwxr-xr-x 1 yoh yoh 23040 Oct 14 14:07 fmap_run+/
drwxr-xr-x 1 yoh yoh 33600 Oct 14 14:09 func_run+_task-1back/
drwxr-xr-x 1 yoh yoh 67200 Oct 14 14:11 func_run+_task-oddball/
drwxr-xr-x 1 yoh yoh 33600 Oct 14 14:08 func_run\=_task-1back/

```

### LINKED ISSUE #4: Smartly handle cases when acq-label is not set for fieldmap
eg

```shell
mvdoc@smaug:/mnt/btrfs/dbic/inbox/DICOM/2016/10/28/A000087$ ls -l
total 0
drwxr-xr-x 1 yoh yoh 14286 Oct 28 13:02 01-anat-scout
drwxr-xr-x 1 yoh yoh   560 Oct 28 13:02 02-anat-scout_MPR_sag
drwxr-xr-x 1 yoh yoh   336 Oct 28 13:02 03-anat-scout_MPR_cor
drwxr-xr-x 1 yoh yoh   336 Oct 28 13:02 04-anat-scout_MPR_tra
drwxr-xr-x 1 yoh yoh 14278 Oct 28 13:03 05-anat-scout
drwxr-xr-x 1 yoh yoh   560 Oct 28 13:03 06-anat-scout_MPR_sag
drwxr-xr-x 1 yoh yoh   336 Oct 28 13:03 07-anat-scout_MPR_cor
drwxr-xr-x 1 yoh yoh   336 Oct 28 13:03 08-anat-scout_MPR_tra
drwxr-xr-x 1 yoh yoh 21468 Oct 28 13:09 09-anat-T1w_acq-MPRAGE
drwxr-xr-x 1 yoh yoh 12510 Oct 28 13:12 10-fmap_run+_2mm
drwxr-xr-x 1 yoh yoh  6256 Oct 28 13:12 11-fmap_run+_2mm
drwxr-xr-x 1 yoh yoh 45248 Oct 28 13:22 12-func_run+_task-faceobject_acq-2mm0.3gap
drwxr-xr-x 1 yoh yoh 44432 Oct 28 13:29 13-func_run+_task-faceobject_acq-2mm0.3gap
drwxr-xr-x 1 yoh yoh 45248 Oct 28 13:36 14-func_run+_task-faceobject_acq-2mm0.3gap
drwxr-xr-x 1 yoh yoh 45248 Oct 28 13:43 15-func_run+_task-faceobject_acq-2mm0.3gap
```

possibly, also handle cases when func following fieldmap is set to `run+` instead of `run=` (iff fmap is not preceded by a func, in that case who knows?)

### LINKED ISSUE #5: allow for separating PI from student with a - (as well as originally planned _)

### PR DESCRIPTION
**Pull request by mvdoc**
Seems to be working. Tomorrow I'll add more. Do not merge yet.

Closes #1 
Closes #4 
Closes #5

### CODE REVIEW

File: heuristics/dbic_bids.py [hunk 1]
```diff
@@ -20,6 +20,38 @@ def create_key(subdir, file_suffix, outtype=('nii.gz', 'dicom'),
     return template, outtype, annotation_classes
 
 
+# XXX: hackhackhack
+protocols2fix = ['dbic^pulse_sequences']
```
Comment (yarikoptic) [inline heuristics/dbic_bids.py]: for those with simple remappings (not requiring any custom logic), you could make it a dict here, from study description to those  replace items.  Then inside you could just check if you have an entry in it, and if so -- use it in that replacement loop

File: heuristics/dbic_bids.py [hunk 2]
```diff
@@ -1,10 +1,60 @@
...
+        [('fmap_run\+_2mm', 'fmap_run+_acq-2mm')]
+}
+keys2replace = ['protocol_name', 'series_description']
+
+
+def filter_files(fn):
+    """Return True if a file should be kept, else False.
+    We're using it to filter out files that do not start with a number."""
+
+    split = fn.split('/')
```
Comment (yarikoptic) [inline heuristics/dbic_bids.py]: In general Use os.path.split fit splitting paths

File: heuristics/dbic_bids.py [hunk 3]
```diff
@@ -1,10 +1,60 @@
...
+         ('DTI_30', 'dwi_run+_acq-30')],
+    '76b36c80231b0afaf509e2d52046e964':
+        [('fmap_run\+_2mm', 'fmap_run+_acq-2mm')]
+}
+keys2replace = ['protocol_name', 'series_description']
+
+
+def filter_files(fn):
+    """Return True if a file should be kept, else False.
+    We're using it to filter out files that do not start with a number."""
```
Comment (yarikoptic) [inline heuristics/dbic_bids.py:108]: Did Terry re-sent for all accession numbers?

File: heuristics/dbic_bids.py [hunk 4]
```diff
@@ -1,10 +1,60 @@
...
+
+    split = fn.split('/')
+    sequence_dir = split[-2]
+
+    return True if re.match('^[0-9]+-', sequence_dir) else False
```
Comment (yarikoptic) [inline heuristics/dbic_bids.py]: Think hope you could avoid here having any if ;-)

File: heuristics/dbic_bids.py [hunk 5]
```diff
@@ -20,6 +70,56 @@ def create_key(subdir, file_suffix, outtype=('nii.gz', 'dicom'),
...
+
+def fix_canceled_runs(seqinfo, accession2run=fix_accession2run):
+    """Function that adds cancelme_ to known bad runs which were forgotten"""
+    accession_number = get_unique(seqinfo, 'accession_number')
+    if accession_number in accession2run:
+        badruns = accession2run[accession_number]
+        badruns_pattern = '|'.join(badruns)
+        for i, s in enumerate(seqinfo):
+            match = re.match(badruns_pattern, s.series_id)
+            if match:
```
Comment (yarikoptic) [inline heuristics/dbic_bids.py]: Minor: so why bother creating match variable and not just use the re.match directly in if?

File: heuristics/dbic_bids.py [hunk 6]
```diff
@@ -20,6 +70,56 @@ def create_key(subdir, file_suffix, outtype=('nii.gz', 'dicom'),
...
+    """Ad-hoc fixup for existing protocols"""
+    # add cancelme to known bad runs
+    seqinfo = fix_canceled_runs(seqinfo)
+
+    # get name of the study to check if we know how to fix it up
+    study_descr = get_unique(seqinfo, 'study_description')
+    study_descr_hash = md5sum(study_descr)
+
+    if study_descr_hash not in subsdict:
+        raise ValueError("I don't know how to fix {0}".format(study_descr))
```
Comment (yarikoptic) [inline heuristics/dbic_bids.py:189]: I thought that if not listed we just don't fix anything

**Review by yarikoptic — COMMENTED** [review]
Great work! Thanks

File: heuristics/dbic_bids.py (hunk 3 above)
Comment (mvdoc) [reply heuristics/dbic_bids.py:108]: it looks so, but we can triple check

File: heuristics/dbic_bids.py (hunk 6 above)
Comment (mvdoc) [reply heuristics/dbic_bids.py:189]: yeah, it shouldn't happen...but who knows? :D

File: heuristics/dbic_bids.py (hunk 3 above)
Comment (mvdoc) [reply heuristics/dbic_bids.py:108]: ok, it wasn't so—see next lines

**Pull request merged**